### PR TITLE
Refactor neutronics

### PR DESCRIPTION
**Pull request by shimwell**
## Proposed changes

This is a refactor for the NeutronicModel. It splits methods into more discrete operations and allows the user to run separate sections for maximum control of the simulation.

NeutronicsModel.simulate has two options that can save time by not remaking files that might already exist
```python
            create_dagmc_geometry: Optional[bool] = True,
            create_openmc_model: Optional[bool] = True,
```

A few more docstrings and type hinting added for good measure :-)

Also the mesh tally location can be specified using the new ```mesh_3d_corners``` and ```mesh_2d_corners``` arguments

Also it is possible to pass a manifest.json file to the NeutronicsModel.geometry argument and skip making the CAD with the Paramak entirely. 

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactoring
- [x] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Pep8 applied
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by shimwell** [comment]
@billingsley-john  when you get time while you are using this new feature anyway are you able to review this?

**Comment by shimwell** [comment]
Tests are now passing locally

**Comment by shimwell** [comment]
looks like some more tests are needed to increase the coverage

**Comment by billingsley-john** [comment]
These are great additions that will save considerable time in some scenarios

**Comment by billingsley-john** [comment]
Both create_dagmc_geometry and create_openmc_model options have been tested and work.

Passing manifest.json straight to NeutronicsModel as the geometry has also been tested and works.

**Review by billingsley-john — APPROVED** [review]
All looks good!

**Comment by shimwell** [comment]
Thanks John. I shall just go through and add some extra tests.

**Comment by shimwell** [comment]
Just needs a few tests that:
- [x] create_dagmc_neutronics_geometry with method set to Trelis and checks for a FileNotFoundError error
- [x] create_dagmc_neutronics_geometry with method set to pymoab and checks for a dagmc.h5
- [x] runs create_openmc_neutronics_model with mesh_3d_corners set to values and with 2d and 3d mesh tallies
- [x] run simulate without any xml files in the folder and with create_openmc_model set to false and checks for a FileNotFoundError
- [x] run simulate without any h5m files in the folder and with create_dagmc_geometry set to false and checks for a FileNotFoundError 
- [x] Use the Shape. export_neutronics_description() with a json filename that doesn't include the .json
- [x] Check the contests of the Shape. export_neutronics_description() contains correct keys and values and is a list of 1 entry

File: paramak/reactor.py [hunk 1]
```diff
@@ -569,15 +567,15 @@ def export_h5m_with_trelis(
                 raise FileNotFoundError("The filename entered as the geometry \
                     argument {} does not exist".format(self.geometry))
```
Comment (shimwell) [inline paramak/reactor.py]:
```suggestion
                    argument {} does not exist".format(self.shapes_and_components))
```

**Pull request merged**